pkg/ccl/changefeedccl/kvevent/blocking_buffer.go

```diff
@@ -39,50 +43,55 @@ type blockingBuffer struct {
 // It will grow the bound account to buffer more messages but will block if it
 // runs out of space. If ever any entry exceeds the allocatable size of the
 // account, an error will be returned when attempting to buffer it.
-func NewMemBuffer(acc mon.BoundAccount, metrics *Metrics, opts ...quotapool.Option) Buffer {
-	bb := &blockingBuffer{
-		signalCh: make(chan struct{}, 1),
-	}
-	bb.acc = acc
-	bb.metrics = metrics
-
+func NewMemBuffer(
+	acc mon.BoundAccount, sv *settings.Values, metrics *Metrics, opts ...quotapool.Option,
+) Buffer {
 	opts = append(opts,
 		quotapool.OnWaitFinish(
 			func(ctx context.Context, poolName string, r quotapool.Request, start time.Time) {
 				metrics.BufferPushbackNanos.Inc(timeutil.Since(start).Nanoseconds())
 			}))
 
-	bb.qp = quotapool.New("changefeed", &bb.blockingBufferQuotaPool, opts...)
-	return bb
+	return &blockingBuffer{
+		signalCh: make(chan struct{}, 1),
+		metrics:  metrics,
+		sv:       sv,
+		qp: allocPool{
+			AbstractPool: quotapool.New("changefeed", &memQuota{acc: acc}, opts...),
+			metrics:      metrics,
+		},
+	}
 }
 
 var _ Buffer = (*blockingBuffer)(nil)
 
-func (b *blockingBuffer) pop() (e *bufferEntry, closed bool) {
+func (b *blockingBuffer) pop() (e *bufferEntry, err error) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
 	if b.mu.closed {
-		return nil, true
+		return nil, io.EOF
 	}
-	return b.mu.queue.dequeue(), false
+	return b.mu.queue.dequeue(), nil
 }
 
 // Get implements kvevent.Reader interface.
 func (b *blockingBuffer) Get(ctx context.Context) (ev Event, err error) {
 	for {
-		got, closed := b.pop()
-		if closed {
-			return Event{}, nil
+		got, err := b.pop()
+		if err != nil {
+			return Event{}, err
 		}
 
 		if got != nil {
+			b.metrics.BufferEntriesOut.Inc(1)
 			e := got.e
 			e.bufferGetTimestamp = timeutil.Now()
-			b.qp.Update(func(r quotapool.Resource) (shouldNotify bool) {
-				res := r.(*blockingBufferQuotaPool)
-				res.release(got)
-				return true
-			})
+			e.alloc = Alloc{
+				bytes:   got.alloc,
+				entries: 1,
+				ap:      b.qp,
```

Comment:
please take the address of `b.qp` here. Otherwise this is going to allocate a new one every time.